src/builder.rs

```diff
@@ -67,14 +69,77 @@ pub enum BuildError {
     CircuitError(#[from] circuit_builder::CircuitBuildError),
 }
 
+#[derive(Default)]
+/// Base builder, can generate builders for containers
+pub struct HugrBuilder {
+    base: HugrMut,
+}
+
+impl HugrBuilder {
+    /// Initialize a new builder
+    pub fn new() -> Self {
+        // initially assume to be a module root, will be replaced if not.
+        Self {
+            base: HugrMut::new_module(),
+        }
+    }
+
+    /// Use this builder to build a module HUGR
+    pub fn module_hugr_builder(&mut self) -> ModuleBuilder {
+        ModuleBuilder(&mut self.base)
+    }
+
+    /// Use this builder to build a DFG HUGR
+    pub fn dfg_hugr_builder(
+        &mut self,
+        input: impl Into<TypeRow>,
+        output: impl Into<TypeRow>,
+    ) -> Result<DFGBuilder, BuildError> {
+        let input = input.into();
+        let output = output.into();
+        let root = self.base.hugr().root();
+        let dfg_op = DataflowOp::DFG {
+            signature: Signature::new_df(input.clone(), output.clone()),
+        };
+        self.base.replace_op(root, dfg_op);
+
+        DFGBuilder::create_with_io(&mut self.base, root, input, output)
+    }
+
+    // TODO: CFG, BasicBlock, Def, Conditional, TailLoop, Case
+
+    /// Complete building and return HUGR, performing validation.
+    pub fn finish(self) -> Result<Hugr, BuildError> {
+        Ok(self.base.finish()?)
+    }
+}
+
+// macro to implement empty drop for builders
+// forces use of value (through calling finish) before automatic drop
```

Comment:
Where does it call `finish`?

Reply:
forces the user to call finish, the previous lines that Alec flagged as "should we call module finish here?" would now throw compiler errors because rust would try to call "drop" at the end of the scope - which requires a mutable reference to be alive. By calling "finish" you ensure the builder leaves scope.

Reply:
It doesn't prevent things like
```rust
let mut builder = HugrBuilder::new();
{
  let mut module_builder = builder.module_hugr_builder();
}
builder.finish()
```
But I guess it avoids some errors.

A more drastic approach would be calling finish or `panic!`ing on the drop, but I'm not sure we want to go there.

Reply:
https://github.com/CQCL-DEV/hugr/issues/115 made an issue here